File: 3721.longest-balanced-subarray-ii.py
#
# @lc app=leetcode id=3721 lang=python3
#
# [3721] Longest Balanced Subarray II
#

# @lc code=start
class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        # HashMap to store first occurrence of balance
        balance_map = {0: -1}
        # Sets to track distinct evens and odds
        even_set = set()
        odd_set = set()
        max_length = 0
        balance = 0  # Difference between counts of evens and odds
        for i, num in enumerate(nums):
            if num % 2 == 0:
                even_set.add(num)
            else:
                odd_set.add(num)
            # Calculate balance as difference between sizes of sets
            balance = len(even_set) - len(odd_set)
            if balance in balance_map:
                max_length = max(max_length, i - balance_map[balance])
            else:
                balance_map[balance] = i  # Store first occurrence of this balance
        return max_length
# @lc code=end

File: 3721.longest-balanced-subarray-ii.cpp
# @lc app=leetcode id=3721 lang=cpp
# [3721] Longest Balanced Subarray II

# @lc code=start
class Solution {
public:
    int longestBalanced(vector<int>& nums) {
        unordered_map<int, int> evenCounts;
        unordered_map<int, int> oddCounts;
        int left = 0;
        int maxLength = 0;
        for (int right = 0; right < nums.size(); ++right) {
            if (nums[right] % 2 == 0) {
                evenCounts[nums[right]]++;
            } else {
                oddCounts[nums[right]]++;
            }
            while (evenCounts.size() != oddCounts.size()) {
                if (nums[left] % 2 == 0) {
                    if (--evenCounts[nums[left]] == 0) {
                        evenCounts.erase(nums[left]);
                    }
                } else {
                    if (--oddCounts[nums[left]] == 0) {
                        oddCounts.erase(nums[left]);
                    }
                }
                ++left;
            }
            maxLength = max(maxLength, right - left + 1);
        }
        return maxLength;
    }
};
# @lc code=end